### PR TITLE
jwm: 2.2.2 -> 1406

### PR DESCRIPTION
**Pull request by romildo**
###### Things done:
- [ ] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- [ ] Built on platform(s): Linux
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
###### More
- [ ] Tested the binary `jwm` using `Xephyr`.
- [ ] Integrated with `services.xserver.windowManager` based on nix expressions for other window managers.
- [ ] Did not test the integration with `services.xserver.windowManager` because I do not know how to test it.

### CODE REVIEW

**Comment by joachifm** [comment]
Would you mind separating the package update from the module addition?

File: pkgs/applications/window-managers/jwm/default.nix [hunk 1]
```diff
@@ -1,22 +1,33 @@
...
+  buildInputs = [
+    pkgconfig automake autoconf libtool libX11 libXext libXinerama libXpm
+    libXft which xorg.libXrender libXau libXdmcp libpng libjpeg expat xproto
+    xextproto xineramaproto librsvg gettext freetype fontconfig
+  ];
 
   preConfigure = ''
-    export NIX_CFLAGS_COMPILE="$NIX_CFLAGS_COMPILE -I${freetype}/include/freetype2 "
-    export NIX_LDFLAGS="$NIX_LDFLAGS -lXft -lfreetype -lfontconfig "
+    ./autogen.sh
```
Comment (hrdinka) [inline pkgs/applications/window-managers/jwm/default.nix]: Have you tried using `autoreconfHook` instead?

File: pkgs/applications/window-managers/jwm/default.nix [hunk 2]
```diff
@@ -1,22 +1,33 @@
...
+     sha256 = "0yk22b7cshhyfpcqnb4p59yxspx95xg9yp1kmkxi2fyw95cacab4";
   };
 
-  buildInputs = [ libX11 libXext libXinerama libXpm libXft freetype 
-    fontconfig ];
+  buildInputs = [
+    pkgconfig automake autoconf libtool libX11 libXext libXinerama libXpm
+    libXft which xorg.libXrender libXau libXdmcp libpng libjpeg expat xproto
+    xextproto xineramaproto librsvg gettext freetype fontconfig
+  ];
```
Comment (hrdinka) [inline pkgs/applications/window-managers/jwm/default.nix]: Some of these are `nativeBuildInputs` (pkgconfig, auto*). `buildInputs` are dependencies needed by the actual program during runtime while `nativeBuildInputs` are needed to build the derivation. For example a x86 linux machine needs native `gcc` to build for arm.

File: pkgs/applications/window-managers/jwm/default.nix [hunk 3]
```diff
@@ -1,22 +1,33 @@
-{ stdenv, fetchurl, libX11, libXext, libXinerama, libXpm, libXft, freetype,
+{ stdenv, fetchurl, pkgconfig, automake, autoconf, libtool, libX11, libXext,
+  libXinerama, libXpm, libXft, which, xorg, libXau, libXdmcp, libpng, libjpeg,
+  expat, xproto, xextproto, xineramaproto, librsvg, gettext, freetype,
   fontconfig }:
 
 stdenv.mkDerivation rec {
-  name = "jwm-2.2.2";
+  name = "jwm-${version}";
+  version = "s1406";
```
Comment (hrdinka) [inline pkgs/applications/window-managers/jwm/default.nix]: A version should not start with an alphabetical character. If it does nix won't be able to distinct between a derivations name and its version.

File: pkgs/applications/window-managers/jwm/default.nix [hunk 4]
```diff
@@ -24,6 +35,8 @@ stdenv.mkDerivation rec {
     '';
 
   meta = {
+    homepage = "http://joewing.net/projects/jwm/";
     description = "A window manager for X11 that requires only Xlib";
+    license = stdenv.lib.licenses.gpl2;
```
Comment (hrdinka) [inline pkgs/applications/window-managers/jwm/default.nix]: The meta attributes lacks a `platforms` and `maintainer` field. Would you like to maintain it? If you aren't in the maintainers list yet you can add yourself in `lib/maintainers.nix`.

File: pkgs/applications/window-managers/jwm/default.nix [hunk 5]
```diff
@@ -1,22 +1,33 @@
...
 
+  configureFlags = [
+    "--enable-shape"
+    "--enable-xrender"
+    "--enable-png"
+  ];
+
   postInstall =
     ''
       sed -i -e s/rxvt/xterm/g $out/etc/system.jwmrc
```
Comment (hrdinka) [inline pkgs/applications/window-managers/jwm/default.nix]: This derivation seems pretty old you might want to check if these patches are still needed.

**Comment by hrdinka** [comment]
> Did not test the integration with services.xserver.windowManager because I do not know how to test it.

@romildo what problems do you have testing it? Can I help you with that or answer any questions?

File: pkgs/applications/window-managers/jwm/default.nix (hunk 1 above)
Comment (romildo) [inline pkgs/applications/window-managers/jwm/default.nix]: Yes. It fails.

File: pkgs/applications/window-managers/jwm/default.nix (hunk 2 above)
Comment (romildo) [inline pkgs/applications/window-managers/jwm/default.nix]: I did not know about `nativeBuildInputs`. I suppose it is relatively new, as most of the nix expressions I used for learning do not use it.

File: pkgs/applications/window-managers/jwm/default.nix (hunk 3 above)
Comment (romildo) [inline pkgs/applications/window-managers/jwm/default.nix]: Then I suppose I should drop the `s` in the version number, although the `.tar.gz` is released with it in the version number.

File: pkgs/applications/window-managers/jwm/default.nix (hunk 4 above)
Comment (romildo) [inline pkgs/applications/window-managers/jwm/default.nix]: Ok.

File: pkgs/applications/window-managers/jwm/default.nix (hunk 5 above)
Comment (romildo) [inline pkgs/applications/window-managers/jwm/default.nix]: They are not.

**Comment by romildo** [comment]
@hrdinka, for a package I would use package `nixpkgs.config.packageOverrides` so that the system configuration could use my version of the nix expression for the package. Would that work for a module?
I have a clone of `nixpkgs` at `/alt/nixpkgs`. The new version for the window manager module is at`/alt/nixpkgs/nixos/modules/services/x11/window-managers/default.nix`. I do not think

    nixpkgs.config.packageOverrides = pkgs: rec {
      window-managers = pkgs.callPackage /alt/nixpkgs/nixos/modules/services/x11/window-managers {};
    }

would work.

So how can I rebuild my system to use my local version of the nix expression so that in the next boot `jwm` will be an option for the display manager?

**Comment by jagajaga** [comment]
LGTM

**Comment by joachifm** [comment]
@romildo The quickest way, I suppose is, to do something like `nixos-rebuild -I nixpkgs=/alt/nixpkgs -I nixos-config=/path/to/configuration-where-jwm-is-enabled.nix build-vm`, which would give you a virtual machine to run. If you're not on NixOS, it becomes slightly more involved as you need to perform the same operations manually.

**Comment by hrdinka** [comment]
@romildo did you have luck testing it? You might want to have a look at the `NIX_PATH` environment variable to set your `nixpkgs` and `nixos-config` permanently. I guess building the new configuration will take while due the switch to the master branch. In the meantime I have built and tested your changes locally and everything works fine. I would be happy to merge this if no one else has something to add.

**Comment by romildo** [comment]
I have tested the changes with `nixos-rebuild boot -I nixpkgs=/alt/nixpkgs --fast` and they worked fine.

**Comment by hrdinka** [comment]
Built (NixOS x86_64) & tested locally. Thanks for your contribution :+1:

**Pull request merged**